docs/module-lizmap/configuration.md
# Configuration

## Dans QGIS

### Configurer le projet QGIS

Pour utiliser le module cadastre dans **Lizmap Web Client** il vous faut d'abord configurer un **projet QGIS** :

* Vous aurez besoin d'installer l'extension QGIS [Cadastre](https://github.com/3liz/QgisCadastrePlugin)
sur votre QGIS Bureautique via le gestionnaire d'extensions, ainsi que sur votre serveur QGIS Serveur
dans un dossier nommé `cadastre`.

* Il faut ensuite créer un projet QGIS contenant les données cadastrales, et le configurer
avec l'extension **Cadastre** :

    - Utilisez l'outil de **chargement des couches**, il configurera notamment les **variables** dans le projet
  ainsi que les couches requises. Sinon, il faut utiliser l'algorithme de QGIS dans le menu **Traitements** pour
  **configurer le projet**.

    ![Propriétés du projet](media/load_data.png)

    - Dans les **propriétés du projet**, onglet `QGIS Serveur`, il faut publier en WFS les couches
  `Communes`, `Parcelles` et `Sections`.

    ![Publication WFS](media/wfs_properties.png)

??? note "Liste des variables qui doivent être dans les propriétés du projet"
    Ces variables sont obligatoires dans les propriétés du projet pour avoir le module Cadastre fonctionnel sur le
    projet en question :

    * `cadastre_commune_layer_id`
    * `cadastre_commune_unique_field`
    * `cadastre_section_layer_id`
    * `cadastre_section_unique_field`
    * `cadastre_parcelle_layer_id`
    * `cadastre_parcelle_unique_field`

### Configurer la publication avec l'extension Lizmap

Vous devez maintenant créer ou modifier la configuration **lizmap** via l'extension Lizmap pour QGIS :

* Il faut obligatoirement ajouter la couche `Parcelles` dans l'onglet **table attributaire**.

    - Pour les performances, pensez à cocher la case **Limiter les données à l'emprise courante** en bas de la fenêtre
      pour l'onglet table attributaire

![Table attributaire](media/table_attributaire.png)

!!! tip Performance
    Pour les performances, vous pouvez aussi cocher **Masquer la couche dans la liste** pour la configuration de la
    couche `Parcelles`. La fonction de sélection sera active, mais la couche `Parcelles` n'apparaîtra pas dans la liste des
    tables attributaires disponibles.

![Configuration Table attributaire de la couche Parcelles](media/table_attributaire_parcelles.png)

* La couche `Parcelles` ne doit pas être dans un groupe qui a l'option `Groupe comme une couche` activée
dans l'onglet `Couches` de Lizmap.

* Vous devez activer la **popup** pour la couche des `Parcelles`, via l'onglet `Couches` de l'extension Lizmap.

!!! warning
    Selon le public visé, vous devrez adapter les champs visibles dans la popup de Lizmap,
    notamment les champs `proprietaire` et `proprietaire_infos` qui contiennent des données sensibles issues
    des fichiers MAJIC.

    - Si vous utilisez la popup **Automatique** de Lizmap, désactiver le WMS pour les champs souhaités dans l'onglet
    `Champs` des propriétés de la couche (tout à droite du tableau)

    - Si vous utilisez la popup **Infobulle QGIS**, adaptez la configuration en supprimant le contenu HTML non souhaité,
    visible dans l'onglet `Affichage` (ou `Infobulle`) des propriétés de la couche, groupe `Infobulle HTML`

* À noter que depuis la version **1.6.0** du module Cadastre, lorsque l'utilisateur a
**le droit de voir les données de propriétaires**, le contenu détaillé des informations de la parcelle est
**affiché sous le contenu de la popup**. Les onglets correspondent à ceux visibles dans QGIS lorsqu'on
affiche le détail d'une parcelle : **Parcelle, Propriétaires, Subdivisions fiscales, Locaux,
Locaux : informations détaillées**. Vous pouvez donc afficher moins d'information dans la popup,
par exemple seulement le code de la parcelle et la commune

## Dans Lizmap Web Client

Maintenant, vous devez configurer les droits d'accès à l'outil via l'interface d'administration.
Le module a créé :

* un groupe d'utilisateurs `Cadastre Lizmap` (identifiant `cadastre_lizmap`)
* 2 sujets de droit : `Accéder aux données de propriétaires` et `Utiliser le panneau de recherche`

Vous pouvez mettre certains utilisateurs dans ce groupe `Cadastre Lizmap`. Ils auront alors accès
au panneau de recherche par localisation **ET** par nom de propriétaires. **Vous devez bien faire
attention de n'accorder le droit `Accéder aux données de propriétaires` qu'aux personnes habilitées !**.

Vous pouvez modifier plus finement les droits via le bouton `Changer les droits des groupes` situés
dans le menu `Groupes d'utilisateurs pour les droits` de l'interface d'administration de Lizmap Web Client.

!!! warning
    Cette page modifie fortement les droits par défaut des groupes d'utilisateurs.
    Nous conseillons de ne pas toucher aux droits liés à Lizmap ici, sauf si vous maîtrisez ce que vous faites.


docs/extension-qgis/configuration.md
# Configuration

Avant d'importer les premières données cadastrales dans la base de données, il faut au préalable configurer
l'extension :

* Menu **Cadastre ➡ Configurer le plugin** ou l'icône **outils** de la barre d'outil Cadastre.

![alt](../media/cadastre_option_dialog.png)

## Interface

Les deux boutons **Interface Cadastre** et **Interface QGIS** permettent d'ouvrir une aide pour appliquer une
interface simplifiée adaptée à une utilisation de consultation du Cadastre.

À ce jour, QGIS ne permet pas de modifier dynamiquement l'interface via une extension. Nous incorporerons cette
fonctionnalité lorsque ce sera possible. En attendant, il faut donc le faire manuellement, comme expliqué dans
la fenêtre d'aide.

## Nom des fichiers MAJIC

Cette partie permet de spécifier comment sont appelés les fichiers MAJIC sur votre poste de travail.

En effet, les conventions de nommage peuvent changer d'un département à l'autre. Souvent, les fichiers se
terminent par une extension relative au département et à la direction, par exemple `.800` pour les fichiers
du département de la Somme.

**Il est important de bien configurer ces noms de fichiers avant votre premier import.**

Si l'extension ne trouve pas les fichiers MAJIC pendant l'import, alors que vous aviez spécifié le bon
répertoire d'import, un message vous avertira et vous proposera d'annuler l'import.

## Modèle de mise en page pour l'export de la vue cartographique

Vous pouvez choisir ici le modèle de mise en page d'impression qui sera utilisé dans la fonction **Export la
vue**. Un modèle est fourni dans le répertoire `composeur` de l'extension et sera utilisé si vous ne proposez
pas le vôtre.

Pour l'instant, le modèle gère un bloc de carte, mais pas les blocs de table attributaire.

## Répertoire temporaire

Vous pouvez choisir le répertoire dans lequel les scripts seront copiés et les fichiers décompressés.

Choisissez un répertoire contenant assez de place pour stocker les fichiers temporaires, surtout si vous
souhaitez charger des données volumineuses.

Ce répertoire est aussi celui dans lequel les **relevés parcellaires** et les **relevés de propriété** seront
exportés.

## Performances

Vous pouvez modifier dans ce groupe les options suivantes pour adapter l'extension aux performances de votre
matériel :

* **Taille maximum des requêtes INSERT** : C'est le nombre total de requêtes `INSERT` lancées dans un groupe de
  modification (`BEGIN/COMMIT`). Vous pouvez baisser le chiffre jusqu'à 10000 si vous avez une machine légère
  et un gros volume de données à importer. Plus le chiffre est bas, plus l'import initial peut prendre du
  temps.

* **Stockage temporaire** : Le mode *MEMORY* est plus rapide, mais nécessite assez de mémoire vive pour
  stocker les données à traiter. Le mode *DEFAULT* est plus lent et adapté à des ordinateurs avec peu de
  mémoire vive.


CHANGELOG.md
# Changelog

## Unreleased

## 1.19.0 - 2023-12-14

* Correction de l'import des données sans les données MAJIC
* Augmentation de la version minimum de QGIS à 3.22 au lieu de 3.18 qui était une non LTR

## 1.18.2 - 2023-10-17

* Utilisation du code commune de la parcelle dans parcelle_info MAJIC
* Correction des DROP d'indexes 
* Correction de la clé primaire de la table proprietaire
* Ajout d'un index sur le champ voie de la table parcelle_info
* Import - Amélioration des performances pour l'EDIGEO
  L'import des données EDIGEO via ogr2ogr n'utilisait plus l'option `PG_USE_COPY`

## 1.18.1 - 2023-09-27

* Corrections SQL

## 1.18.0 - 2023-08-02

* Support du MAJIC 2023 (contribution de @landryb et @MaelREBOUX)

## 1.17.3 - 2023-02-07

* bugfix - La colonne RC TEOM au tableau des propriétés baties est le champ `bateom` et pas `baeteom`

## 1.17.2 - 2023-01-26

* bugfix - La colonne RC TEOM au tableau des propriétés baties est le champ `baeteom` et pas `mvltieomx` de `pevtaxation`
* bugfix - homogénéisation des adresses de propriétaires
* Tableau des indivisions comme sous partie du tableau des propriétaires
* Indivisions pour un tiers

## 1.17.1 - 2022-12-15

* bugfix - Correction du compte communal abrégé affiché dans les relevés
* Ajout de la colonne RC TEOM au tableau des propriétés baties
* bugfix - Server - Accès à la fiche HTML complète sans les infos sensibles des propriétaires

## 1.17.0 - 2022-11-25

* Server - Ajout de la carte au relevé parcellaire
* Server - Support du relevé parcellaire pour un tiers et réduction de la fiche HTML

## 1.16.1 - 2022-10-19

* Correction d'un bug si le dossier de stockage temporaire dans les préférences n'existait plus

## 1.16.0 - 2022-09-05

* Export des relevés - Ajout d'une option "Les relevés sont destinés à des tiers"
  qui permet de ne pas faire figurer la date et le lieu de naissance des propriétaires dans les PDF.
* Annulation d'une fonctionnalité de la version 1.14.0 concernant l'ajout automatique des noms courts.
  Il y a un problème sur Lizmap Web Client https://github.com/3liz/lizmap-web-client/issues/2985

## 1.15.1 - 2022-08-12

* Identification de la parcelle (MAJIC): ajout du statut 'Parcelle bâtie' dans la fiche (issue du champ `gparbat`).
  Cette modification ne nécessite pas de réimport pour être fonctionnelle.
* Table `parcelle_info`: ajout du champ `parcelle_batie` (Oui/Non) à partir du champ `gparbat` (MAJIC).
  Cette modification nécessite un réimport des données, ou un remplacement manuel de la table `parcelle_info`.

## 1.15.0 - 2022-08-02

* Support du MAJIC 2022

## 1.14.2 - 2022-05-10

* Correction de l'import Edigeo si les parcelles sont trop grandes (en Guyane par exemple, ticket https://github.com/3liz/QgisCadastrePlugin/issues/206)
* Correction d'une erreur Python si l'utilisateur utilise QGIS Process

## 1.14.1 - 2022-03-21

* Correction d'une régression côté serveur pour le logging

## 1.14.0 - 2022-03-09

* Ajout de la colonne `contenance` à la table `parcelle_info` même lorsqu'il n'y a pas de données MAJIC
* Correction de la recherche par propriétaire lorsque le schéma Postgres nécessite des doubles quotes
* Ajout d'une clé primaire à la table `geo_label`
* Modification de la clé primaire de la table `geo_batiment` et `geo_subdsect` pour être de type entier
* Ajout des noms courts sur les couches et les groupes pour rendre le projet valide par défaut lors d'une publication
  sur Lizmap
* Refactorisation du code côté QGIS Serveur
* Mise à jour de https://docs.3liz.org

## 1.13.5 - 2021-12-14

* Amélioration du message d'erreur sur le fichier n'est pas être décompressé, ticket #339
* Correction de la table `commune_majic` pour les millésimes 2019 à 2021, ticket #333

## 1.13.4 - 2021-12-06

* Ajout du financeur de la métropole de Clermont-Ferrand
* Correction d'une régression suite à la version 1.13.3 sur les champs `*_bipevla`

## 1.13.3 - 2021-11-18

* Corrections orthographiques
* Correction sur les "Locaux : Informations détaillées"
* Correction des exports PDF depuis le panneau de recherche
* Amélioration du message d'erreur s'il n'y a pas assez de mémoire sur l'ordinateur
* Amélioration de la documentation concernant les paramètres au niveau global de l'extension

## 1.13.2 - 2021-08-25

* Mise à jour de https://docs.3liz.org/QgisCadastrePlugin/ pour les financeurs de l'extension
* Correction d'une erreur si l'extension est utilisée pour la première fois
* Correction d'une erreur possible pour ouvrir les fichiers en UTF-8 par défaut

## 1.13.1 - 2021-08-10

* Correction d'un chemin dans le script standalone concernant l'export

## 1.13.0 - 2021-08-10

* Ajout du support MAJIC 2021 (Contribution de Maël Reboux)
* Refactorisation du code interne à l'extension
* Mise à jour de la documentation https://docs.3liz.org/QgisCadastrePlugin/

## 1.12.1 - 2021-06-24

* Correction du chargement de l'extension avec QGIS < 3.10

## 1.12.0 - 2021-06-21

* Correction de l'export propriétaire à une commune
* Mise à jour de la documentation sur le module Lizmap

## 1.11.1 - 2021-06-01

* Sauvegarde et restoration des paramètres de la dernière base utilisée dans l'interface graphique

## 1.11.0 - 2021-05-21

* Correction d'un espace dans l'adresse dans les propriétés baties, non baties
* Correction d'une erreur SQL lors de la consultation des infos parcelle sur QGIS 3.16 bureautique et serveur
* Ajout du support QGIS 3.16
* Ajout d'un bouton pour ouvrir l'aide depuis le menu principal d'aide de QGIS
* Ajout de la recherche d'un propriétaire
  * par nom d'usage ou de naissance
  * par ville
* Ajout d'un téléchargeur Edigéo communal dans la boite à outils Processing
* Documentation https://docs.3liz.org/QgisCadastrePlugin/
  * Mise de la documentation au format MkDocs
  * Intégration de la documentation du module Lizmap
  * Ajout de la documentation automatique concernant les traitements Processing
  * Ajout de la documentation concernant la base de données avec SchémaSpy
* Ajout d'une infrastructure de tests


cadastre/processing/algorithms/config.py
"""
/***************************************************************************
 Cadastre - Processing config algorithms
                                 A QGIS plugin
 This plugins helps users to import the french land registry ('cadastre')
 into a database. It is meant to ease the use of the data in QGIs
 by providing search tools and appropriate layer symbology.
                              -------------------
        begin                : 2019-05-15
        copyright            : (C) 2019 by 3liz
        email                : info@3liz.com
 ***************************************************************************/

/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/
"""

from qgis.core import (
    QgsProcessing,
    QgsProcessingOutputNumber,
    QgsProcessingParameterField,
    QgsProcessingParameterVectorLayer,
)

from cadastre.definitions import URL_DOCUMENTATION
from cadastre.processing.algorithms.base import BaseProcessingAlgorithm


class ConfigProjectAlgorithm(BaseProcessingAlgorithm):
    """
    Algorithm to set project variables for cadastre use
    """

    COMMUNE_LAYER = 'COMMUNE_LAYER'
    COMMUNE_UNIQUE_FIELD = 'COMMUNE_UNIQUE_FIELD'
    SECTION_LAYER = 'SECTION_LAYER'
    SECTION_UNIQUE_FIELD = 'SECTION_UNIQUE_FIELD'
    PARCELLE_LAYER = 'PARCELLE_LAYER'
    PARCELLE_UNIQUE_FIELD = 'PARCELLE_UNIQUE_FIELD'

    SUCCESS = 'SUCCESS'

    def initAlgorithm(self, config):
        self.addParameter(
            QgsProcessingParameterVectorLayer(
                self.COMMUNE_LAYER,
                self.tr('La couche communes'),
                [QgsProcessing.TypeVectorPolygon],
                defaultValue='Communes'
            )
        )

        self.addParameter(
            QgsProcessingParameterField(
                self.COMMUNE_UNIQUE_FIELD,
                self.tr('Champs identifiant les communes'),
                parentLayerParameterName=self.COMMUNE_LAYER,
                defaultValue='geo_commune',
                type=QgsProcessingParameterField.String
            )
        )

        self.addParameter(
            QgsProcessingParameterVectorLayer(
                self.SECTION_LAYER,
                self.tr('La couche sections'),
                [QgsProcessing.TypeVectorPolygon],
                defaultValue='Sections',
            )
        )

        self.addParameter(
            QgsProcessingParameterField(
                self.SECTION_UNIQUE_FIELD,
                self.tr('Champs identifiant les sections'),
                parentLayerParameterName=self.SECTION_LAYER,
                defaultValue='geo_section',
                type=QgsProcessingParameterField.String
            )
        )

        self.addParameter(
            QgsProcessingParameterVectorLayer(
                self.PARCELLE_LAYER,
                self.tr('La couche parcelles'),
                [QgsProcessing.TypeVectorPolygon],
                defaultValue='Parcelles',
            )
        )

        self.addParameter(
            QgsProcessingParameterField(
                self.PARCELLE_UNIQUE_FIELD,
                self.tr('Champs identifiant les parcelles'),
                parentLayerParameterName=self.PARCELLE_LAYER,
                defaultValue='geo_parcelle',
                type=QgsProcessingParameterField.String
            )
        )

        self.addOutput(QgsProcessingOutputNumber(self.SUCCESS, self.tr('Succès')))

    def processAlgorithm(self, parameters, context, feedback):
        commune_layer = self.parameterAsVectorLayer(parameters, self.COMMUNE_LAYER, context)
        commune_unique_field = self.parameterAsString(parameters, self.COMMUNE_UNIQUE_FIELD, context)

        section_layer = self.parameterAsVectorLayer(parameters, self.SECTION_LAYER, context)
        section_unique_field = self.parameterAsString(parameters, self.SECTION_UNIQUE_FIELD, context)

        parcelle_layer = self.parameterAsVectorLayer(parameters, self.PARCELLE_LAYER, context)
        parcelle_unique_field = self.parameterAsString(parameters, self.PARCELLE_UNIQUE_FIELD, context)

        variables = context.project().customVariables()

        variables['cadastre_commune_layer_id'] = commune_layer.id()
        variables['cadastre_commune_unique_field'] = commune_unique_field

        variables['cadastre_section_layer_id'] = section_layer.id()
        variables['cadastre_section_unique_field'] = section_unique_field

        variables['cadastre_parcelle_layer_id'] = parcelle_layer.id()
        variables['cadastre_parcelle_unique_field'] = parcelle_unique_field

        context.project().setCustomVariables(variables)
        # Returns empty dict if no outputs
        return {self.SUCCESS: 1}

    def name(self):
        return 'config_project'

    def displayName(self):
        return self.tr('Configuration du projet')

    def shortHelpString(self):
        return (
            'Ce traitement permet de configurer un projet QGIS pour de la publication sur le web avec '
            '<a href="https://github.com/3liz/lizmap-web-client/">Lizmap Web Client</a> et son '
            '<a href="https://github.com/3liz/lizmap-cadastre-module">module Cadastre</a>.\n'
        )

    def helpUrl(self) -> str:
        return f"{URL_DOCUMENTATION}/processing/#configuration-du-projet"
